7659 7661 kode forbedringer

ragnarwestad opened the pull request:
Konverterte Steg-klassesystemet til TypeScript (MELOSYS-7659, MELOSYS-7661)

  ## Sammendrag

  Konverterte kjernekomponentene i steg-systemet fra JavaScript til TypeScript:
  - Base-klassen `Steg` (steg.js → steg.ts)
  - To steg-implementasjoner: `Periode` og `Trygdeavgift`
  - Oppdaterte type-definisjoner i `typer.ts`
  - Fikset ustabile snapshot-tester

  Alle endringer er **100% bakoverkompatible** med eksisterende 40+ steg-klasser som arver fra Steg.

  ## Filer endret

  ### Konvertert til TypeScript
  - `src/felleskomponenter/stegvelger/stegMotor/steg.js` → `steg.ts` (base-klasse)
  - `src/sider/eu_eøs/saksbehandling/stegMap/periode.js` → `periode.ts`
  - `src/sider/eu_eøs/saksbehandling/stegMap/trygdeavgift.js` → `trygdeavgift.ts`

  ### Oppdaterte filer
  - `src/felleskomponenter/stegvelger/stegMotor/typer.ts` - Korrekte typer for avklartefakta/vilkar
  - `src/sider/eu_eøs/saksbehandling/stegMap/periode.test.ts` - Nye unit tests (20 tester)
  - `src/sider/eu_eøs/saksbehandling/stegMap/trygdeavgift.test.ts` - Nye unit tests (17 tester)
  - `src/sider/eu_eøs/stegKomponenter/vurderingPeriode/vurderingPeriode.test.tsx` - Fikset dato-avhengige snapshots

  ## Tekniske detaljer

  ### Type-forbedringer

  **Steg.ts (base-klasse):**
  - Alle felter har eksplisitte typer med `protected` visibility
  - Return-typer definert på alle metoder
  - Bruker `!` (non-null assertions) for å bevare identisk runtime-oppførsel
  - Type-imports fra typer.ts

  **PropsLight og StegKriterie:**
  ```typescript
  // Korrigert fra Record<string, unknown> til unknown[]
  avklartefakta?: unknown[];
  vilkar?: unknown[];

  Periode og Trygdeavgift:
  - Fjernet StegWithTypes hack - bruker nå direkte this-referanser
  - Index signatures på interfaces for base-klasse kompatibilitet
  - Eksplisitte return-typer på alle funksjoner

  Test-stabilitet

  vurderingPeriode snapshot-tester:
  - Mocker systemtid til fast dato (15. juni 2024)
  - Forhindrer daglige feil pga. endring av "today" i datovelger
  - Bruker vi.setSystemTime() i beforeAll og vi.useRealTimers() i afterAll

  Testing og verifisering

  ✅ 37 nye unit tests (20 periode + 17 trygdeavgift) - alle passerer✅ 27 vurderingPeriode tests - alle passerer (inkl. 4 snapshot-tester)✅ TypeScript kompilerer uten feil✅ Build vellykket i produksjonsmodus✅ Git-historikk bevart via git mv før TypeScript-konvertering

  Runtime-oppførsel

  100% bakoverkompatibel:
  - Identisk runtime-oppførsel som original JavaScript
  - Fail-fast ved manglende felter (kræsjer hvis ikke satt)
  - Alle 40+ eksisterende steg-klasser fungerer uendret
  - Ingen breaking changes

  Fremtidige muligheter

  Nå som base-klassen er TypeScript:
  - De resterende ~40 steg-klassene kan konverteres gradvis
  - Kan fjerne type-hacks og workarounds i child-klasser
  - Bedre type-sikkerhet ved utvikling av nye steg

ragnarwestad committed:
7659 7661 Legg til testdekning for MELOSYS-7659 og MELOSYS-7661 (Fase 1)

  Etablerer baseline behavior for periode- og trygdeavgift-stegene
  før TypeScript-konvertering og refaktorering. Følger TDD-approach
  fra prosjektets KODESTANDARD.md.

  Nye testfiler:
  - periode.test.ts (20 tester)
    * Constructor med ART11_3A → ART11_5 mapping
    * samleRelevanteData og beregnRelevantUI funksjoner
    * Handlers og neste steg logikk

  - trygdeavgift.test.ts (17 tester)
    * Constructor og behandlingID-håndtering
    * Handlers og no-op oppdaterStatus
    * Alltid harAvklaring: true

  - vurderingPeriode.test.tsx (27 tester)
    * Rendering med ulike props
    * Form submission og validering
    * Periode-forkortning interaksjoner
    * useEffect behaviors
    * Edge cases og snapshot testing

  Oppdaterte filer:
  - Stegvelger.test.tsx (2 nye test-cases)
    * Verifiserer VURDERING_PERIODE enum
    * Verifiserer VURDERING_TRYGDEAVGIFT enum

  Total testdekning: 66 nye tester
  Alle tester passerer: ✅

ragnarwestad committed:
7659 7661 For typer.ts:
  TODO-26 Steg 5: Konverterte typer.js til TypeScript

  La til 'as const' på FANE_STATUS og STEG for literal types, eksporterte FaneStatus og StegType union types, og la til PropsLight og StegKriterie interfaces.

ragnarwestad committed:
7659 7661 Konverterte periode.js til TypeScript

  Brukte StegWithTypes interface pattern for å type JS base class properties.
  Reduserte fra 13 any-casts til 1 kontrollert cast i constructor.
  Oppdaterte periode.test.ts med type casts for å håndtere JS base class
  getters/setters.

ragnarwestad committed:
7659 7661 Oppdaterte periode.test.ts med type casts

  La til type casts for å håndtere JS base class getters/setters:
  - Cast samleRelevanteData til any før kall
  - Cast beregnRelevantUI til any før kall
  - Cast kriterier til any for array-tilgang
  - Cast handlers til any i beforeEach for gjenbruk i tester

ragnarwestad committed:
7659 7661 Oppdaterte trygdeavgift.test.ts med type casts

  La til type casts for å håndtere JS base class getters/setters:
  - Cast samleRelevanteData til any før kall
  - Cast beregnRelevantUI til any før kall
  - Cast kriterier til any for array-tilgang
  - Cast handlers til any i beforeEach for gjenbruk i tester
  - La til manglende byggLovvalgsperioder mock i createMockPropsLight

ragnarwestad committed:
7659 7661 Konverterte trygdeavgift.js til TypeScript

  Brukte StegWithTypes interface pattern for å type JS base class properties.
  La til type annotations for PropsLight og StegKriterie imports.
  Typet constructor-parametere og handler-funksjoner med eksplisitte typer.

ragnarwestad committed:
7659 7661 Konverterte index.js til TypeScript

  La til type annotation for stegMap: Map<string, any> med norske kommentarer.
  Bruker 'any' siden ulike steg-klasser har forskjellige metode-signaturer utover
  basis Steg-klassen (f.eks. SaksbehandlingInngang har ekstra metoder).

  Type assertion (as any) nødvendig på Map-konstruktøren for å håndtere heterogene
  constructor-typer.

ragnarwestad committed:
7659 7661 Erstatt Record<string, unknown> med spesifikke interfaces i steg-klasser

  Forbedrer type-sikkerhet i periode.ts og trygdeavgift.ts ved å erstatte
  generiske Record-typer med spesifikke interfaces:

  - periode.ts: PeriodeRelevanteData, PeriodeRelevantUI, PeriodeHandlers
  - trygdeavgift.ts: TrygdeavgiftRelevanteData, TrygdeavgiftRelevantUI, TrygdeavgiftHandlers

  Dette gir bedre autofullføring og type-sjekking, og gjør koden mer vedlikeholdbar.

ragnarwestad committed:
7659 7661 Konverterte Steg.js til TypeScript og fjernet StegWithTypes-hack

  - Konverterte felleskomponenter/stegvelger/stegMotor/steg.js til TypeScript
  - La til TypeScript-typer for alle properties, getters/setters og metoder
  - Fjernet StegWithTypes-hack fra periode.ts og trygdeavgift.ts
  - Reduserte antall 'any'-typer fra 3 til 1 (kun for React.ComponentType)

  Index signatures (midlertidig løsning):
  La til `[key: string]: unknown;` på steg-interfacene (PeriodeRelevanteData,
  PeriodeRelevantUI, PeriodeHandlers, osv.) for å gjøre dem kompatible med
  base-klassens `Record<string, unknown>`-krav. Dette gir kompatibilitet med
  base-klassen samtidig som vi beholder type safety på navngitte properties.

  Dette er en midlertidig løsning som kan fjernes når alle 44 subklasser av Steg
  er konvertert til TypeScript. Da kan vi gjøre Steg generisk og få full type
  safety uten index signatures. Nå ville det kreve å endre 42 JavaScript-filer
  som vi ikke ønsker å røre i dette prosjektet.

  Type improvements:
  - Byttet `(relevantUI as any).harAvklaring` til type-safe `"harAvklaring" in relevantUI`
  - Beholdt `React.ComponentType<any>` med eslint-disable da base-klassen må
    akseptere komponenter med ulike Props-typer (alternativet ville vært generics
    som ville krevd endringer i alle 44 filer som arver fra Steg

ragnarwestad committed:
Revert "7659 7661 Konverterte Steg.js til TypeScript og fjernet StegWithTypes-hack"

This reverts commit fdede2a4c65d409aff58942c9b9c0192fca5b5ed.

ragnarwestad committed:
7659 7661 Korrigeringer etter QA fra CoPilot.

- Fikset historikk på steg.ts
- Satt avklartefakta og vilkar til arrays.
- Korrigeret dato-håndtering i vurderingPeriode.test.tsx

"
  nesteSteg = (): string | undefined => {
    const { avklartefakta = {}, vilkar = {} } = this._propsLight;

The destructuring default values were changed from arrays ([]) to objects ({}). While this aligns with the TypeScript types (Record<string, unknown>), this changes runtime behavior. If existing code elsewhere relies on these being arrays, this could cause issues. Verify that avklartefakta and vilkar are indeed always objects and never arrays in the codebase.
"

ragnarwestad committed:
7659 7661 Fjernet optional chaining fra Steg for identisk runtime-oppførsel

  Fjernet defensive programming (optional chaining) fra steg.ts for å
  sikre 100% identisk runtime-oppførsel med original JavaScript.

  **Endringer:**
  - Byttet `?.` til `!` (non-null assertions) i byggSteg(), nesteSteg() og hentStatus()
  - Tilbake til `kriterieMatch && kriterieMatch.nesteSteg` i stedet for `kriterieMatch?.nesteSteg`
  - Fjernet type guard, tilbake til `relevantUI.harAvklaring`

  **Resultat:**
  - Identisk fail-fast oppførsel som original JS
  - Kræsjer hvis felter ikke er satt (i stedet for silent failure)
  - Type-sikkerhet beholdt, men ingen endring i runtime-logikk
  - Alle 37 tester passerer
  - Build vellykket

CK-Hui commented on src/sider/eu_eøs/saksbehandling/stegMap/periode.test.ts:
```diff
@@ -0,0 +1,285 @@
+      const periode = new Periode(propsLight, 5);
+
+      const samleRelevanteData = periode.samleRelevanteData as any;
+      const relevantData = samleRelevanteData(propsLight);
```
Klarer ikke å se at `samleRelevanteData` blir brukt noen steder?
Er ikke den det samme som `relevanteData`?

ragnarwestad replied:
Den har setter/getter'e på ca linje 90 og brukes i byggSteg på linje 110.
Jeg ser ingen relevanteData?

Det står "Outdated" over kommentaren her, mulig vi blir lurt :-)

CK-Hui commented on src/sider/eu_eøs/saksbehandling/stegMap/trygdeavgift.test.ts:
```diff
@@ -0,0 +1,240 @@
+    });
+  });
+
+  describe("hentStatus", () => {
```
Må innrømme at jeg ikke skjenner til Fane status så godt, men om den er knyttet til om man kan gå videre i steget så vil ikke den alltid være OK. Men kanskje jeg blander litt her. 

Det er egentlig litt bug i det vi har i dag, 
- trygdeavgift skal ikke være aktiv før kravene for å gå videre er oppfylt i periode steget.
- Vedtak steget skal ikke være aktiv før trygdeavgift har oppfylt steg kravene. Det vil si:
   - Brukere har svart på "ja" - "nei"
   - Hvis "ja" så kan man bare gå videre
   - Hvis "nei" så må man beregne trygdeavgift først, fylt inn inntekt- og skatteperioder!

CK-Hui commented on src/sider/eu_eøs/saksbehandling/stegMap/trygdeavgift.ts:
```diff
@@ -0,0 +1,64 @@
+interface TrygdeavgiftRelevanteData {
+  [key: string]: unknown; // Index signature for base class compatibility
+  behandlingID?: string;
+  redigerbart: boolean;
```
Trenger å legge til `aktivSteg` prop

ragnarwestad replied:
Ser ikke noe aktivSteg heller jeg.

ragnarwestad committed:
7659 7661 Fjernet ubrukte handler-interfaces fra periode og trygdeavgift

  Ryddet opp i kode som var igjen etter fjerning av StegWithTypes-hacken:

  **Endringer:**
  - Fjernet PeriodeHandlers interface (periode.ts) - ikke i bruk
  - Fjernet TrygdeavgiftHandlers interface (trygdeavgift.ts) - ikke i bruk
  - Fjernet ubrukt import av StegKriterie fra begge filer
  - Fjernet unødvendig eslint-disable kommentar i vurderingPeriode.tsx

  **Bakgrunn:**
  Disse interfaces ble opprettet i commit a10bebb91 og brukt via
  StegWithTypes-hacken. Da hacken ble fjernet i commit 2f2bf63b0,
  ble interfacene liggende igjen uten å være i bruk, siden base class
  Steg bruker Record<string, unknown> for handlers (nødvendig for
  kompatibilitet med 37 JavaScript-filer som ikke er konvertert ennå).
